src/impl/collection_change_builder.cpp

```diff
@@ -652,3 +653,137 @@ CollectionChangeSet CollectionChangeBuilder::finalize() &&
         std::move(columns)
     };
 }
+
+void ObjectChangeSet::insertions_add(ObjectKeyType obj)
+{
+    m_insertions.insert(obj);
+}
+
+void ObjectChangeSet::modifications_add(ObjectKeyType obj, ColKeyType col)
+{
+    auto it = m_modifications.find(obj);
+    if (it != m_modifications.end()) {
+        it->second.insert(col);
+    }
+    else {
+        m_modifications.insert({obj, {col}});
+    }
+}
+
+void ObjectChangeSet::deletions_add(ObjectKeyType obj)
+{
+    m_modifications.erase(obj);
+    size_t num_inserts_removed = m_insertions.erase(obj);
+    if (num_inserts_removed == 0) {
+        m_deletions.insert(obj);
+    }
+}
+
+void ObjectChangeSet::clear(size_t old_size)
+{
+    static_cast<void>(old_size); // unused
+    m_clear_did_occur = true;
+    m_insertions.clear();
+    m_modifications.clear();
+    m_deletions.clear();
+}
+
+bool ObjectChangeSet::insertions_remove(ObjectKeyType obj)
+{
+    return m_insertions.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::modifications_remove(ObjectKeyType obj)
+{
+    return m_modifications.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::deletions_remove(ObjectKeyType obj)
+{
+    return m_deletions.erase(obj) > 0;
+}
+
+bool ObjectChangeSet::deletions_contains(ObjectKeyType obj) const
+{
+    if (m_clear_did_occur) {
+        // FIXME: what are the expected notifications when an object is deleted
+        // and then another object is inserted with the same key?
+        return m_insertions.count(obj) == 0;
+    }
+    return m_deletions.count(obj) > 0;
+}
+
+bool ObjectChangeSet::insertions_contains(ObjectKeyType obj) const
+{
+    return m_insertions.count(obj) > 0;
+}
+
+bool ObjectChangeSet::modifications_contains(ObjectKeyType obj) const
+{
+    return m_modifications.count(obj) > 0;
+}
+
+util::Optional<std::pair<ObjectChangeSet::ColKeyIterator, ObjectChangeSet::ColKeyIterator>> ObjectChangeSet::get_columns_modified(ObjectKeyType obj) const
```

Comment:
This could just return a const pointer to `it->second` which would make the calling code less awkward.

Reply:
Thanks, I changed to using a pointer, which is much simpler.